### DISCUSSION TITLE
Could you please help me for investigate installation steps on tomcat?

### DISCUSSION THREAD

**Discussion by KwanchaiJc** [apiman/apiman#2147] [Q&A] [answered]
I installed apiman on tomcat already but i can't do anything on apimanui
Please provide full steps install on tomcat or help me investigate this error. I'm newbie.

I've disabled es on apiman.properties already because i've found error : Unable to reach Elasticsearch (with error: connection refused) Retry will be attempted in 5 seconds.
![1656412071797](https://user-images.githubusercontent.com/85068934/176355919-bd1d6a88-474d-4f43-9306-0cd644e1ae6a.jpeg)

Server information
![image](https://user-images.githubusercontent.com/85068934/176355455-abc6c53e-0ab7-41ed-9f71-90a47214a054.png)
![image](https://user-images.githubusercontent.com/85068934/176355488-0ab35434-a06c-4aba-ac8d-3939ed1bd9ce.png)
![image](https://user-images.githubusercontent.com/85068934/176355534-e236a80d-016f-4102-ad81-084d629ded52.png)
Apache Tomcat/9.0.64
Apiman version 2.2.3.Final

Errors:
![176183986-caf87b6a-6b8c-471d-864c-8e4c180a4968](https://user-images.githubusercontent.com/85068934/176355637-6b6f061f-7ffc-4c07-9b8c-531c8c8a52b0.png)
![176183865-7717daab-83db-48bc-b7de-b056dabf4ce5](https://user-images.githubusercontent.com/85068934/176355640-c9a9b45b-4a7f-427d-a681-90796b2bfdca.png)
![176184503-32fd5a07-12ed-40a7-9c43-655e095eab45](https://user-images.githubusercontent.com/85068934/176355642-f18ec96b-9697-4deb-9ba9-5db387724302.png)

APIMAN Properties
[apimanproperties.zip](https://github.com/apiman/apiman/files/9007392/apimanproperties.zip)

**Comment by msavy** [comment]
Hi,

Can you describe what you're trying to do? Are you trying to use a RDBMS?

**Reply by KwanchaiJc** [reply]
Hi @msavy ,
i want to install apiman on tomcat that it. but i used apiman.properties by default (using es) it's show many error that why i disabled es in apiman.properties to other source.

**Comment by KwanchaiJc** [comment, marked as answer]
For answer, see: https://github.com/orgs/apiman/discussions/2147#discussioncomment-3176380

---

Which steps i miss?
1.sudo useradd -m -U -d /opt/tomcat -s /bin/false tomcat
2.wget https://dlcdn.apache.org/tomcat/tomcat-9/v9.0.64/bin/apache-tomcat-9.0.64.tar.gz
3.sudo tar xzvf apache-tomcat-9*tar.gz -C /opt/tomcat --strip-components=1
4.sudo chown -R tomcat:tomcat /opt/tomcat
5.sudo sh -c 'chmod +x /opt/tomcat/bin/*.sh'
find the java location
6.readlink -f $(which java)
e.g. /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.332.b09-1.el7_9.x86_64
7.sudo vi /etc/systemd/system/tomcat.service

Insert these statement & Change JAVA_HOME

[Unit]
Description=Apache Tomcat Web Application Container
After=network.target

[Service]
Type=oneshot
RemainAfterExit=yes

User=tomcat
Group=tomcat

Environment="JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.312.b07-1.el7_9.x86_64/jre"
Environment="JAVA_OPTS=-Djava.security.egd=file:///dev/urandom -Djava.awt.headless=true"
Environment="CATALINA_BASE=/opt/tomcat"
Environment="CATALINA_HOME=/opt/tomcat"
Environment="CATALINA_PID=/opt/tomcat/temp/tomcat.pid"
Environment="CATALINA_OPTS=-Xms512M -Xmx1024M -server -XX:+UseParallelGC"
ExecStart=/opt/tomcat/bin/startup.sh
ExecStop=/opt/tomcat/bin/shutdown.sh

[Install]
WantedBy=multi-user.target


8.sudo systemctl daemon-reload
9.sudo systemctl enable tomcat
10.sudo systemctl start tomcat
11.sudo systemctl status tomcat
12.firewall-cmd --zone=public --permanent --add-port=8080/tcp
13.firewall-cmd --reload
14.curl -L https://github.com/apiman/apiman/releases/download/2.2.3.Final/apiman-distro-tomcat-2.2.3.Final-overlay.zip -o apiman-distro-tomcat8-2.2.3.Final-overlay.zip
15.unzip -o apiman-distro-tomcat8-2.2.3.Final-overlay.zip -d tomcat/
16.sudo chown -R tomcat:tomcat /opt/tomcat
17.sudo sh -c 'chmod +x /opt/tomcat/bin/*.sh'
18.vi /opt/tomcat/conf/tomcat-users.xml

add these statement 

  <role rolename="apiuser"/>
  <role rolename="apiadmin"/>
  <role rolename="apipublisher"/>
  <role rolename="admin-gui"/>
  <role rolename="manager-gui"/>
  <role rolename="manager-script"/>
  <role rolename="manager-jmx"/>
  <role rolename="manager-status"/>
  <user username="admin" password="admin123!" roles="apiuser,apiadmin,admin-gui,manager-gui,manager-script,manager-jmx,manager-status"/>
  <user username="apimanager" password="apiman123!" roles="apipublisher,admin-gui,manager-gui,manager-script,manager-jmx,manager-status"/>
 

19.sudo vi /opt/tomcat/webapps/manager/META-INF/context.xml
comment ""




20.sudo vi /opt/tomcat/webapps/host-manager/META-INF/context.xml
comment ""




21.sudo systemctl restart tomcat
22.sudo systemctl stop tomcat


if i using default apiman.perproties, Have i to install elasticsearch ? or Can i change to other source?

**Reply by msavy** [reply]
You don't have to use Elasticsearch, but you will need to modify `apiman.properties` to change the quickstart default; you can have it use  whatever you like (e.g. all kinds of different SQL database). Check out the docs as there are implementations of all the components in JDBC/SQL (including metrics, rate limiting, etc).

**Comment by msavy** [comment]
Hi,

You need to run the Elasticsearch database on port 9200 and point Apiman to
it. Should be the default.

Or you need to set up a relational database and alter a few areas of Apiman
and Tomcat config (see docs). Other community members might chime in with
more detail, as I'm traveling right now.

There are some easy docker images for Elasticsearch, or you can run ES as a
Linux service.

If you're inexperienced, it might be easier to start with the wildfly
distro.

**Reply by msavy** [reply]
@KwanchaiJc Can you please explain what you did so that other members of the community can benefit from your experience and the time spent by the team to help you?

**Reply by KwanchaiJc** [reply]
Hi everyone

1.sudo useradd -m -U -d /opt/tomcat -s /bin/false tomcat
2.wget https://dlcdn.apache.org/tomcat/tomcat-8/v8.5.81/src/apache-tomcat-8.5.81-src.tar.gz
3.sudo tar xzvf apache-tomcat-9*tar.gz -C /opt/tomcat --strip-components=1
4.sudo chown -R tomcat:tomcat /opt/tomcat
5.sudo sh -c 'chmod +x /opt/tomcat/bin/*.sh'
find the java location
6.readlink -f $(which java)
e.g. /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.332.b09-1.el7_9.x86_64
7.sudo vi /etc/systemd/system/tomcat.service

Insert these statement & Change JAVA_HOME

```
[Unit]
Description=Apache Tomcat Web Application Container
After=network.target

[Service]
Type=oneshot
RemainAfterExit=yes

User=tomcat
Group=tomcat

Environment="JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.312.b07-1.el7_9.x86_64/jre"
Environment="JAVA_OPTS=-Djava.security.egd=file:///dev/urandom -Djava.awt.headless=true"
Environment="CATALINA_BASE=/opt/tomcat"
Environment="CATALINA_HOME=/opt/tomcat"
Environment="CATALINA_PID=/opt/tomcat/temp/tomcat.pid"
Environment="CATALINA_OPTS=-Xms512M -Xmx1024M -server -XX:+UseParallelGC"
ExecStart=/opt/tomcat/bin/startup.sh
ExecStop=/opt/tomcat/bin/shutdown.sh

[Install]
WantedBy=multi-user.target
```

8.sudo systemctl daemon-reload
9.sudo systemctl enable tomcat
10.sudo systemctl start tomcat
11.sudo systemctl status tomcat
12.firewall-cmd --zone=public --permanent --add-port=8080/tcp
13.firewall-cmd --reload
14.curl -L https://github.com/apiman/apiman/releases/download/2.2.3.Final/apiman-distro-tomcat-2.2.3.Final-overlay.zip -o apiman-distro-tomcat8-2.2.3.Final-overlay.zip
15.unzip -o apiman-distro-tomcat8-2.2.3.Final-overlay.zip -d tomcat/
16.sudo chown -R tomcat:tomcat /opt/tomcat
17.sudo sh -c 'chmod +x /opt/tomcat/bin/*.sh'
18.vi /opt/tomcat/conf/tomcat-users.xml then add these statement

```
  <role rolename="apiuser"/>
  <role rolename="apiadmin"/>
  <role rolename="apipublisher"/>
  <role rolename="admin-gui"/>
  <role rolename="manager-gui"/>
  <role rolename="manager-script"/>
  <role rolename="manager-jmx"/>
  <role rolename="manager-status"/>
  <user username="admin" password="admin123!" roles="apiuser,apiadmin,admin-gui,manager-gui,manager-script,manager-jmx,manager-status"/>
  <user username="apimanager" password="apiman123!" roles="apipublisher,admin-gui,manager-gui,manager-script,manager-jmx,manager-status"/>
```

19.sudo vi /opt/tomcat/webapps/manager/META-INF/context.xml
comment ""
```
<!--
<Valve className="org.apache.catalina.valves.RemoteAddrValve"
   allow="127\.\d+\.\d+\.\d+|::1|0:0:0:0:0:0:0:1" />
-->
```

20.sudo vi /opt/tomcat/webapps/host-manager/META-INF/context.xml
comment ""
```
<!--
<Valve className="org.apache.catalina.valves.RemoteAddrValve"
   allow="127\.\d+\.\d+\.\d+|::1|0:0:0:0:0:0:0:1" />
-->
```

21.sudo systemctl restart tomcat

APIMAN and Mysql Integration Steps
1.Create Database apiman in Mysql
2.Execure mysql ddls
 2.1 Go to "/opt/tomcat/apiman/ddls/" path
 2.2 Download and Execute apiman-gateway_mysql5.ddl
 2.3 Download and Execute apiman_mysql5.ddl
3.Download JDBC Driver to "/opt/tomcat/lib"
4.Stop tomcat service
5.Edit apiman.properties file
 5.1 Disable all es parameters
 5.2 Enable other parameter Ex. apiman.properties -> [apiman.zip](https://github.com/apiman/apiman/files/9136934/apiman.zip)
 5.3 Edit context.xml Ex. context.xml
Add
```
<Resource name="jdbc/datasources/apiman-manager" auth="Container" 
    type="javax.sql.DataSource"
                   maxTotal="100" maxIdle="30" maxWaitMillis="10000"
                   username="Your_Username" password="Your_Password" 
    driverClassName="com.mysql.jdbc.Driver"
                   url="jdbc:mysql://Your_Mysql_IP:3306/apiman"/>
         <ResourceLink global="jdbc/datasources/apiman-manager" 
    name="jdbc/datasources/apiman-manager" type="javax.sql.DataSource"/>
    <Resource name="jdbc/datasources/apiman-gateway" auth="Container"
    type="javax.sql.DataSource"
                   maxTotal="100" maxIdle="30" maxWaitMillis="10000"
                   username="Your_Username" password="Your_Password"
                   driverClassName="com.mysql.jdbc.Driver"
                   url="jdbc:mysql://Your_Mysql_IP:3306/apiman"/>
         <ResourceLink global="jdbc/datasources/apiman-gateway"
    name="jdbc/datasources/apiman-gateway" type="javax.sql.DataSource"/>
```

 5.4 Edit web.xml Ex. web.xml
Add
```
<resource-ref>
        <res-ref-name>jdbc/datasources/apiman-manager</res-ref-name>
        <res-type>javax.sql.DataSource</res-type>
        <res-auth>Container</res-auth>
    </resource-ref>
```

6.Start tomcat service

**Reply by msavy** [reply]
Well done, thank you for sharing.
